### PR TITLE
improve s3 key manger, add key selection logics

### PR DESCRIPTION
**Pull request by lizk886**
1 improve s3 key manger for unit testing
2 get a more realistic key selection logic

### CODE REVIEW

File: src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java [hunk 1]
```diff
@@ -6,13 +6,9 @@
...
 import io.vertx.core.json.JsonObject;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.ArrayList;
-import java.util.Set;
-import java.util.HashSet;
+
+import java.util.*;
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java:10]: Do we need star import? Lets avoid this.

File: src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java [hunk 2]
```diff
@@ -84,21 +82,22 @@ public Collection<S3Key> getKeysForSite(Integer siteId) {
...
-                        largestKey = key;
-                    }
-                }
-            }
-            return largestKey;
         }
+        return findLargestKey(keys);
     }
+
+    S3Key findLargestKey(Collection<S3Key> keys) {
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java:97]: Do we use this anywhere besides the above function? Seems like we can just get rid of it

File: src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java [hunk 3]
```diff
@@ -84,21 +82,22 @@ public Collection<S3Key> getKeysForSite(Integer siteId) {
                 .collect(Collectors.toList());
     }
 
+    public List<S3Key> getKeysForSiteFromMap(int siteId) {
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java:85]: I thinks getKeys is a better name

File: src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java (hunk 2 above)
Comment (lizk886) [reply src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java:103]: `findLargestKey` is currently being used by `getEncryptionKeyForSite`. Right now, I am using a temporary key selection method by simply choosing the largest key for a specific siteId. I planned to use this temporary key selection logic until someone started work on key rotation.

File: src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java [hunk 4]
```diff
@@ -23,7 +28,7 @@ public class RotatingS3KeyProvider implements StoreReader<Map<Integer, S3Key>> {
     ScopedStoreReader<Map<Integer, S3Key>> reader;
 
     private static final Logger LOGGER = LoggerFactory.getLogger(RotatingS3KeyProvider.class);
-    public final Map<Integer, List<S3Key>> siteToKeysMap = new HashMap<>();
+    Map<Integer, List<S3Key>> siteToKeysMap = new ConcurrentHashMap<>();
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java:31]: Why is this changed to a concurrentHashMap?

File: src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java [hunk 5]
```diff
@@ -84,21 +93,31 @@ public Collection<S3Key> getKeysForSite(Integer siteId) {
                 .collect(Collectors.toList());
     }
 
-    public S3Key getEncryptionKeyForSite(Integer siteId) {
-        Collection<S3Key> keys = getKeysForSite(siteId);
+    public List<S3Key> getKeys(int siteId) {
+        return siteToKeysMap.getOrDefault(siteId, new ArrayList<>());
+    }
+
+    public List<S3Key> getMostRecentKeysForSite(int siteId) {
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java:100]: getActiveKeys

File: src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java [hunk 6]
```diff
@@ -84,21 +93,31 @@ public Collection<S3Key> getKeysForSite(Integer siteId) {
...
+            throw new IllegalStateException("No keys found for site ID: " + siteId);
+        }
+        return keys.stream()
+                .sorted(Comparator.comparingLong(S3Key::getCreated).reversed())
+                // Get most recently created 3 keys
+                .limit(3)
+                .collect(Collectors.toList());
+    }
+
+     public S3Key getEncryptionKeyForSite(Integer siteId) {
```
Comment (cody-constine-ttd) [inline src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java:112]: I think we can combine this with the function above, and always send all active keys.

File: src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java (hunk 4 above)
Comment (lizk886) [reply src/main/java/com/uid2/shared/store/reader/RotatingS3KeyProvider.java]: I am a little bit worried that there would be a concurrency issue,  such as simultaneous updates to the siteId map keys and reading keys occurring at the same time. Can revert this back to hashMap if thats not a problem

**Review by cody-constine-ttd — APPROVED** [review]
